Migrate almost all v1 quirks to v2

puddly opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->
This PR migrates _almost_ all straightforward V1 quirks to V2. This was mostly done with an automated tool to hint LLM agents about the exact quirk `signature` -> `replacement` transformation. This covers about half, the other half are:
- Most of Aqara/Xiaomi
- All of Tuya
- XBee

I don't expect this PR to be merged as-is (we could...). We can use bits and pieces of it and migrate individual manufacturers (or groups of manufacturers) at once.

Some nice APIs to have in the future:
1. Regex or wildcard `applies_to()`, for King of Fans.
2. I'd like to move away from modifying the underlying ZCL device. We can use an overlay to change metadata, if possible:
   - `replaces_endpoint` -> `modify_device_type(endpoint_id=1, device_type=zha.DeviceType.DIMMABLE_LIGHT)`?
   - `.replaces(DoublingPowerConfig1CRCluster, endpoint_id=1)` can likely be replaced with a hint to the underlying `PowerConfiguration` cluster entity. Entity creation hints might be be a good middle ground API between quirk IDs and directly creating entities from quirks.
   - All instances of `.removes()` should probably be replaced with `.prevent_default_entity_creation()`
3. The only uses of `filter` thus far are `.filter(lambda device: WWAH_CLUSTER_ID not in device.endpoints[1].in_clusters)`, in lieu of using firmware version filters.
4. Data bus objects: needs more thought.
5. `_CONSTANT_ATTRIBUTES` can also be replaced with a hint to the underlying ZCL cluster entity? We really only override color mode, mireds, and feature bitmaps.

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->


## Device diagnostics
<!--
  Diagnostics data is used by ZHA unit tests to make sure quirks create the expected
  entities and we do not have regressions. For all new quirks, we require device
  diagnostics.

  You can find the diagnostics information by going to the device page, clicking the
  three dots, and then by clicking on "Download diagnostics". Drag-and-drop the
  downloaded file into this section.
-->
Too many and not enough.


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [ ] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works
- [ ] Device diagnostics data has been attached

puddly committed:
Migrate Centralite quirks

puddly committed:
Migrate quirks, pass 1

puddly committed:
Migrate quirks, pass 2

puddly committed:
Migrate quirks, pass 3

puddly committed:
Migrate quirks, pass 4

puddly committed:
Exclude Samjin multi sensor from Centralite quirk
The Samjin quirk technically could be merged with the Centralite one
once the cluster removal is given more attention.

puddly committed:
Migrate quirks, pass 5

puddly committed:
Migrate Aqara curtain motor

puddly committed:
Migrate quirks, pass 6 (IKEA)

puddly committed:
Migrate quirks, pass 7

puddly committed:
Migrate quirks, pass 9

puddly committed:
Migrate quirks, pass 8

puddly committed:
Migrate quirks, pass 10

puddly committed:
Migrate quirks, pass 11

puddly committed:
Migrate quirks, pass 12

puddly committed:
Migrate quirks, pass 13

puddly committed:
Migrate quirks, pass 14

puddly committed:
Migrate quirks, pass 15

puddly committed:
Migrate quirks, pass 16

puddly committed:
Migrate quirks, pass 17

puddly committed:
Migrate quirks, pass 18

puddly committed:
Migrate quirks, pass 19

puddly committed:
Migrate quirks, pass 20

puddly committed:
Migrate quirks, pass 21

puddly committed:
Migrate quirks, pass 22

puddly committed:
Migrate quirks, pass 23

puddly committed:
Update KOF quirk with device info from ZHA database

puddly committed:
Fix typo in Konke quirk

puddly committed:
Fix Aqara E1 driver

puddly committed:
More fixes

puddly committed:
Use v2 device class for Xiaomi

puddly committed:
Migrate quirks, pass 24

puddly committed:
Migrate quirks, pass 25

puddly committed:
Migrate quirks, pass 26

puddly committed:
WIP: Fix most failing tests

puddly committed:
Fix up bad quirks

puddly committed:
Ignore custom quirk loading test

puddly committed:
Fix KOF quirks

puddly committed:
Minimize diff

puddly committed:
Pre-commit

puddly committed:
Bump zigpy to get rid of direction warning

puddly committed:
Regenerate `uv.lock`

TheJulianJES commented on zhaquirks/zhongxing/motion.py:
```diff
@@ -3,76 +3,20 @@
-
-    def __init__(self, *args, **kwargs):
-        """Init."""
-        self.occupancy_bus = Bus()
```
This is unfortunately still used by `MotionCluster`, so we can't just remove that.

puddly committed:
Manual clean up

puddly committed:
Merge branch 'dev' into puddly/quirks-v2-migration-test2

puddly committed:
Fixes after merge

puddly committed:
One more check

puddly commented:
I've made a few more passes and can't spot any more issues, this might be good to go 😄.
  - `samjin/multi2.py`: Deleted, merged into `smartthings/multi.py` under both possible manufacturer names
  - `hivehome/mot003V6.py`: Deleted, was a duplicate of `mot003V0.py`
  - `centralite/ias.py`: Merged into `centralite/cl_3300S.py`
  - `osram/tunablewhite.py`: Added missing `OsramLightCluster`

A few didn't use model or manufacturer names for filtering. Those I matched up with public info from ZHA diagnostics or Z2M info.

TheJulianJES commented on zhaquirks/zbeacon/doorsensor.py:
```diff
@@ -1,71 +1,15 @@
+# The model zbeacon DS01 sounds a lot like ("eWeLink", "DS01") from Sonoff sold as Sonoff SNZB-04
+# The device tested is sold as Elivco and IHseno IH-MC01 and uses a TuYa ZTU module as described here:
+#     https://github.com/dresden-elektronik/deconz-rest-plugin/issues/7415
+(QuirkBuilder("zbeacon", "DS01").removes(PollControl, endpoint_id=1).add_to_registry())
```
side note: We sometimes used `# fmt: skip` to not get this auto-formatted if the quirk is very small.
See this example here: [thirdreality/temperature_sensor.py#L70-L74](https://github.com/zigpy/zha-device-handlers/blob/9b5774d6cd76e624b377fcba4200c6b4e39a4889/zhaquirks/thirdreality/temperature_sensor.py#L70-L74)

E.g. this should be:
```suggestion
(
    QuirkBuilder("zbeacon", "DS01")
    .removes(PollControl, endpoint_id=1)
    .add_to_registry()
)  # fmt: skip
```
Also fine to leave as is though.

puddly committed:
Avoid one-line quirks

puddly committed:
Merge branch 'dev' into puddly/quirks-v2-migration-test2

puddly committed:
Use spaces for indentation...

puddly committed:
Address review comments

puddly commented on zhaquirks/xiaomi/__init__.py:
```diff
@@ -122,28 +123,15 @@ def __init__(self, *args, **kwargs):
+        if not hasattr(self, BATTERY_SIZE):
+            self.battery_size = BatterySize.CR2032
+        super().__init__(*args, **kwargs)
 
 
```
No longer necessary, since we have reverted the zigpy changed that warranted this workaround.

puddly commented on zhaquirks/elko/smart_super_thermostat.py:
```diff
@@ -108,9 +100,15 @@ async def write_attributes(self, attributes, manufacturer=None):
-            self.endpoint.device.thermostat_bus.listener_event(
-                "heating_active_change", value
-            )
+            if value == 0:
+                mode = self.RunningMode.Off
+                state = self.RunningState.Idle
+            else:
+                mode = self.RunningMode.Heat
+                state = self.RunningState.Heat_State_On
+
+            self._update_attribute(self.attributes_by_name["running_mode"].id, mode)
+            self._update_attribute(self.attributes_by_name["running_state"].id, state)
```
This shifts code around, there isn't a functional change.

puddly committed:
Fix changed uv.lock and pyproject.toml

puddly commented on zhaquirks/terncy/pp01.py:
```diff
@@ -47,56 +27,23 @@ def __init__(self, *args, **kwargs):
-    device_automation_triggers = BUTTON_TRIGGERS
+(
+    QuirkBuilder("Xiaoyan", "TERNCY-PP01")
+    .applies_to("Terncy", "TERNCY-PP01")
```
V2 quirks have no way to match "no manufacturer name". This is the only other variant for the device that I could find.

TheJulianJES commented on zhaquirks/elko/__init__.py:
`heating_active_change` was moved to `zhaquirks/elko/smart_super_thermostat.py`, so I think we can remove this entirely (as well as `ElkoThermostat.thermostat_bus`)?

puddly replied:
I think the whole base class can go.

puddly committed:
Remove unused base `ElkoThermostatCluster`

puddly committed:
Fix formatting and remove unnecessary `cluster_id=`

puddly committed:
Merge branch 'dev' into puddly/quirks-v2-migration-test2

TheJulianJES commented on zhaquirks/kof/kof_mr101z.py:
```diff
@@ -76,43 +65,20 @@ class KofLevelControl(NoReplyMixin, CustomCluster, LevelControl):
+    QuirkBuilder("King Of Fans,  Inc.", "MR101Z")
+    # TODO: add support for wildcard matching in `applies_to`
+    .applies_to("King Of Fans,  Inc.", "HDC52EastwindFan")
+    .applies_to("King Of Fans,  Inc.", "HBUniversalCFRemote")
```
We may wanna check Z2M for all possible models. EDIT: Not finding more there from a quick search.

puddly commented:
@TheJulianJES Merging this PR unfortunately breaks ZHA CI in an annoying way, since it explicitly tests a few devices from quirks by directly importing the device objects. There's some deeper issue elsewhere and creating an equivalent device for quirks to match at runtime seems to change something subtly in tests. In the interest of time, let's postpone this merge.

TheJulianJES commented:
Thanks for checking before merging 😄

puddly committed:
Migrate almost all v1 quirks to v2

Squashed from puddly/quirks-v2-migration-test2 and rebased onto dev.

puddly committed:
Fix no-op `.removes()` calls passing cluster classes instead of IDs

`QuirkBuilder.removes()` takes a cluster ID; passing the cluster class
made the removal silently do nothing. Found by regenerating ZHA device
diagnostics against this branch.

puddly committed:
Restore matching behavior lost in v2 migration

- RODRET: re-add `RODRET wireless dimmer` model and Groups cluster removal
- Konke buttons: restore `KONKE_BUTTON` quirk_id via a CustomDeviceV2 subclass
- Heiman CO_V15/SMOK_YDLV10/CO_CTPG: match "Heiman", not "HEIMAN"

Found by diffing registered (manufacturer, model) match pairs against dev
and regenerating ZHA device diagnostics.

TheJulianJES commented:
Just tried the label auto-trigger for quirks here. Maybe not the best PR to test with 😅😅
